In [1]:
from google.colab import files
uploaded = files.upload()

Saving loan_data.csv to loan_data.csv


In [2]:
import pandas as pd
import io

for loan_data in uploaded.keys():
    print(f"Uploaded file: {loan_data}")
    df = pd.read_csv(io.BytesIO(uploaded[loan_data]))
    print(df.head())

Uploaded file: loan_data.csv
   person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49               

In [3]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings( "ignore" )

* Analyze the given dataset on Colab using pandas and matplotlib (if required). Include:
		○ The shape of the dataset.
		○ The unique values of the output variable.
		○ A short paragraph in text form describing the dataset (included in the Colab notebook).

In [4]:
# The shape of the dataset.
df.shape

(45000, 14)

In [5]:
# The unique values of the output variable.
Y = df['loan_status']
Y.value_counts()

loan_status
0    35000
1    10000
Name: count, dtype: int64

In [6]:
# A short paragraph in text form describing the dataset (included in the Colab notebook).
df.describe()

person_age  person_income  person_emp_exp     loan_amnt  \
count  45000.000000   4.500000e+04    45000.000000  45000.000000   
mean      27.764178   8.031905e+04        5.410333   9583.157556   
std        6.045108   8.042250e+04        6.063532   6314.886691   
min       20.000000   8.000000e+03        0.000000    500.000000   
25%       24.000000   4.720400e+04        1.000000   5000.000000   
50%       26.000000   6.704800e+04        4.000000   8000.000000   
75%       30.000000   9.578925e+04        8.000000  12237.250000   
max      144.000000   7.200766e+06      125.000000  35000.000000   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
count   45000.000000         45000.000000                45000.000000   
mean       11.006606             0.139725                    5.867489   
std         2.978808             0.087212                    3.879702   
min         5.420000             0.000000                    2.000000   
25%         8.590000             0.070000                    3.000000   
50%        11.010000             0.120000                    4.000000   
75%        12.990000             0.190000                    8.000000   
max        20.000000             0.660000                   30.000000   

       credit_score   loan_status  
count  45000.000000  45000.000000  
mean     632.608756      0.222222  
std       50.435865      0.415744  
min      390.000000      0.000000  
25%      601.000000      0.000000  
50%      640.000000      0.000000  
75%      670.000000      0.000000  
max      850.000000      1.000000

• Preprocess the data:
		○ Drop irrelevant columns.
		○ Handle NaN values appropriately.

In [7]:
# Drop irrelevant columns
X = df.drop(columns=['person_age','person_gender','person_education','person_home_ownership','loan_intent','previous_loan_defaults_on_file'])

In [8]:
#Handle NaN values appropriately.
X[X.eq("NaN").any(axis=1)]
#Since there is no NaN values, no need to drop any row

Empty DataFrame
Columns: [person_income, person_emp_exp, loan_amnt, loan_int_rate, loan_percent_income, cb_person_cred_hist_length, credit_score, loan_status]
Index: []

In [13]:
# Split the dataset into training and testing sets.
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler() # max value = 1 , min value = 0
X = scaler.fit_transform(X)

from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
Y = encoder.fit_transform(Y)

X_train,X_test,Y_train,Y_test=train_test_split(X,Y,train_size=0.9, shuffle=True,random_state=69)

#Initially SMOTE was not added
# from imblearn.over_sampling import SMOTE
# smot=SMOTE(random_state=42)
# X_train,Y_train=smot.fit_resample(X_train,Y_train)

• Create two models:
		1. One using the imported Logistic Regression model.
		2. Another by implementing a Logistic Regression model as outlined in the provided resources.

In [14]:
from sklearn.linear_model import LogisticRegression

# Our model based on class and resources

class LogitRegression() :
    def __init__( self, learning_rate, iterations ) :
        self.learning_rate = learning_rate
        self.iterations = iterations

    # Function for model training
    def fit( self, X, Y ) :
        # no_of_training_examples, no_of_features
        self.m, self.n = X.shape
        # weight initialization
        self.W = np.zeros( self.n )
        self.b = 0
        self.X = X
        self.Y = Y

        # gradient descent learning

        for i in range( self.iterations ) :
            self.update_weights()
        return self

    # Helper function to update weights in gradient descent

    def update_weights( self ) :
        A = 1 / ( 1 + np.exp( - ( self.X.dot( self.W ) + self.b ) ) )

        # calculate gradients
        tmp = ( A - self.Y.T )
        tmp = np.reshape( tmp, self.m )
        dW = np.dot( self.X.T, tmp ) / self.m
        db = np.sum( tmp ) / self.m

        # update weights
        self.W = self.W - self.learning_rate * dW
        self.b = self.b - self.learning_rate * db

        return self

    # Hypothetical function  h( x )

    def predict( self, X ) :
        Z = 1 / ( 1 + np.exp( - ( X.dot( self.W ) + self.b ) ) )
        Y = np.where( Z > 0.5, 1, 0 )
        return Y



In [15]:
# Training the sklearn model

# Model training
model = LogitRegression( learning_rate = 0.01, iterations = 1000 )

model.fit( X_train, Y_train )
model1 = LogisticRegression()
model1.fit( X_train, Y_train)

# Prediction on test set
Y_pred = model.predict( X_test )
Y_pred1 = model1.predict( X_test )

# measure performance
correctly_classified = 0
correctly_classified1 = 0

# counter
count = 0
for count in range( np.size( Y_pred ) ) :

    if Y_test[count] == Y_pred[count] :
      correctly_classified = correctly_classified + 1

    if Y_test[count] == Y_pred1[count] :
        correctly_classified1 = correctly_classified1 + 1

    count = count + 1

In [16]:
# Checking Accuarcy and comparing them

print( "Accuracy on test set by our model       :  ", (
  correctly_classified / count ) * 100 )
# In line 43 SMOTE functions changes the outcome of accuarcy for our model.
print( "Accuracy on test set by sklearn model   :  ", (
  correctly_classified1 / count ) * 100 )

Accuracy on test set by our model       :   79.24444444444444
Accuracy on test set by sklearn model   :   100.0


Some explanation by my side: Initially when I trained my model without SMOTE the accuaracy was 79.244 percent. Then when i decieded to add SMOTE which added synthetic data inorder to balance the dataframe, the accuracy for both the model was 100%.

Conclusion: SMOTE is a great way to improve accuracy.